### DISCUSSION TITLE
RabbitMQ 3.12.6 API returns incorrect json result

### DISCUSSION THREAD

**Discussion by stevenxi** [rabbitmq/rabbitmq-server#10156] [Other] [answered]
### Describe the bug

The API '/api/queues' returns queue item array [], instead of expected json.

Expected:
```
{
filtered_count: 1,
item_count: 1,
items: [{arguments:{}...},{arguments:{}...},{arguments:{}...}],
page: 1,
page_count: 1,
page_size: 100,
total_count: 1
}
```

Actual:
```
[
{arguments:{}...},
{arguments:{}...},
{arguments:{}...},
]
```

As a result, the management web portal throws error:

```
TypeError: Cannot read properties of undefined (reading 'length') TypeError: Cannot read properties of undefined (reading 'length') at Array.process (eval at compile (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/ejs-1.0.min.js:1:6654), :7:19) at EJS.render (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/ejs-1.0.min.js:1:1885) at format (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1201:21) at https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:488:24 at with_reqs (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1183:9) at https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1179:17 at req.onreadystatechange (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1260:17)
```

![image](https://github.com/rabbitmq/rabbitmq-server/assets/3340119/bb8b7800-e116-4a93-afb9-81e381f852e8)

### Reproduction steps

1. get result from /api/queues
2.  It returns
```
[
{arguments:{}...},
{arguments:{}...},
{arguments:{}...},
]
```
Or, open the management portal, it shows error
```
TypeError: Cannot read properties of undefined (reading 'length') TypeError: Cannot read properties of undefined (reading 'length') at Array.process (eval at compile (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/ejs-1.0.min.js:1:6654), :7:19) at EJS.render (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/ejs-1.0.min.js:1:1885) at format (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1201:21) at https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:488:24 at with_reqs (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1183:9) at https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1179:17 at req.onreadystatechange (https://tech-rabbitmq.udintel.eu/rabbit-jpijgeqei/js/main.js:1260:17)
```

### Expected behavior

```
{
filtered_count: 1,
item_count: 1,
items: [{arguments:{}...},{arguments:{}...},{arguments:{}...}],
page: 1,
page_count: 1,
page_size: 100,
total_count: 1
}
```

**Comment by michaelklishin** [comment]
Without a specific `curl` (or similar) command and a set of other steps to reproduce, we cannot suggest much. `{arguments:{}...}` is not a value I can add to a test case or run through a JSON serializer.

It could be an Erlang-specific JSON serialization problem where an empty proplist (a list of pairs) and a regular list (a JSON array) are indistinguishable, so an empty JSON object is serialized as an empty list. We have to specifically handled that in multiple places. However, we cannot tell with most of data elided in the examples.

It also matters if the queues have emitted all stats at least once, or they have just been created and most stats are simply not available yet.

So, please provide a specific set of steps we can use to reproduce. It is also important to know if this happens just on the first request or two, or is persistent.

**Reply by stevenxi** [reply]
hi @michaelklishin ,

I don't even know how to get to make it generate the invalid json ***from a fresh environment***.

The API endpoint instead  returning the correct json, with count/paging info and an array of queue objects, it returns the queue objects at the message's root level.

I've tried to reproduce it from a non-production environment, but failed.

Here's one from a message from test virtual host of production enviroment:

<details><summary>JSON</summary>

```
[
    {
        "arguments": {
            "x-queue-type": "classic"
        },
        "auto_delete": true,
        "backing_queue_status": {
            "avg_ack_egress_rate": 0.5692389747442522,
            "avg_ack_ingress_rate": 0.5692389747442522,
            "avg_egress_rate": 0.5692389747442522,
            "avg_ingress_rate": 0.5692389747442522,
            "delta": [
                "delta",
                "undefined",
                0,
                0,
                "undefined"
            ],
            "len": 0,
            "mode": "default",
            "next_deliver_seq_id": 243,
            "next_seq_id": 243,
            "num_pending_acks": 0,
            "num_unconfirmed": 0,
            "q1": 0,
            "q2": 0,
            "q3": 0,
            "q4": 0,
            "qs_buffer_size": 0,
            "target_ram_count": "infinity",
            "version": 1
        },
        "consumer_capacity": 1.0,
        "consumer_utilisation": 1.0,
        "consumers": 1,
        "durable": false,
        "effective_policy_definition": {
            "ha-mode": "all"
        },
        "exclusive": true,
        "exclusive_consumer_tag": null,
        "garbage_collection": {
            "fullsweep_after": 65535,
            "max_heap_size": 0,
            "min_bin_vheap_size": 46422,
            "min_heap_size": 233,
            "minor_gcs": 87
        },
        "head_message_timestamp": null,
        "memory": 65072,
        "message_bytes": 0,
        "message_bytes_paged_out": 0,
        "message_bytes_persistent": 0,
        "message_bytes_ram": 0,
        "message_bytes_ready": 0,
        "message_bytes_unacknowledged": 0,
        "message_stats": {
            "ack": 247,
            "ack_details": {
                "rate": 0.6
            },
            "deliver": 247,
            "deliver_details": {
                "rate": 0.6
            },
            "deliver_get": 247,
            "deliver_get_details": {
                "rate": 0.6
            },
            "deliver_no_ack": 0,
            "deliver_no_ack_details": {
                "rate": 0.0
            },
            "get": 0,
            "get_details": {
                "rate": 0.0
            },
            "get_empty": 0,
            "get_empty_details": {
                "rate": 0.0
            },
            "get_no_ack": 0,
            "get_no_ack_details": {
                "rate": 0.0
            },
            "publish": 247,
            "publish_details": {
                "rate": 0.6
            },
            "redeliver": 0,
            "redeliver_details": {
                "rate": 0.0
            }
        },
        "messages": 0,
        "messages_details": {
            "rate": 0.0
        },
        "messages_paged_out": 0,
        "messages_persistent": 0,
        "messages_ram": 0,
        "messages_ready": 0,
        "messages_ready_details": {
            "rate": 0.0
        },
        "messages_ready_ram": 0,
        "messages_unacknowledged": 0,
        "messages_unacknowledged_details": {
            "rate": 0.0
        },
        "messages_unacknowledged_ram": 0,
        "name": "AD:0585749b-ad81-4204-b9e9-8c9e62cca0d4",
        "node": "rabbit@rabbitmq-UDIntel-Cluster-Linux-Sys-1",
        "operator_policy": null,
        "policy": "ha-all",
        "recoverable_slaves": null,
        "reductions": 185951,
        "reductions_details": {
            "rate": 391.4
        },
        "single_active_consumer_tag": null,
        "state": "running",
        "type": "classic",
        "vhost": "TestHost"
    }
]
```
</details>

**Reply by lukebakken** [reply]
@stevenxi please take a look at how I used `<details>` to hide the large amount of text you pasted behind a dropdown. This makes following the discussion here _MUCH_ easier.

**Reply by stevenxi** [reply]
Thanks  @lukebakken ,

That's actually a simplified version, the original json was around 1MB.

**Comment by stevenxi** [comment]
Just to confirm, issue presists even if move to [3.13.0-rc.2-management](https://hub.docker.com/layers/library/rabbitmq/3.13.0-rc.2-management/images/sha256-a6388040701297c2be13c37d1f5a3369f2782d33386a8ae6d9d02a2c31bc10f3?context=explore)

**Comment by michaelklishin** [comment]
We cannot address what we do not know how to reproduce even some of the
time.

Does rabbitmqctl list_queues —formatter=pretty_table display various
columns differently from other clusters?

Can a cluster definitions file and a filtered query HTTP API response be
shared? You can see what the response is in browser developer tools.

Don’t forget to edit out the credentials in both the definitions file and
any API request examples

On Fri, 15 Dec 2023 at 11:56, Steven.Xi ***@***.***> wrote:

> Just to confirm, issue presists even if move to 3.13.0-rc.2-management
> <https://hub.docker.com/layers/library/rabbitmq/3.13.0-rc.2-management/images/sha256-a6388040701297c2be13c37d1f5a3369f2782d33386a8ae6d9d02a2c31bc10f3?context=explore>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/10156#discussioncomment-7866179>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAIQV62AUKJ7QAVTA6RFLYJR6M5AVCNFSM6AAAAABAWME22KVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQNRWGE3TS>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>

-- 
This electronic communication and the information and any files transmitted 
with it, or attached to it, are confidential and are intended solely for 
the use of the individual or entity to whom it is addressed and may contain 
information that is confidential, legally privileged, protected by privacy 
laws, or otherwise restricted from disclosure to anyone else. If you are 
not the intended recipient or the person responsible for delivering the 
e-mail to the intended recipient, you are hereby notified that any use, 
copying, distributing, dissemination, forwarding, printing, or copying of 
this e-mail is strictly prohibited. If you received this e-mail in error, 
please return the e-mail to the sender, delete it from your computer, and 
destroy any printed copy of it.

**Comment by lukebakken** [comment, marked as answer]
@stevenxi if you can reproduce this in a test environment, feel free to zip up RabbitMQ's data directories and attach them here. That should allow us to easily reproduce the issue.

**Comment by stevenxi** [comment]
Sorry guys, really can't reproduce in a testing environment.

We decided no to use the management portal.
Thank you for looking into it though...

Closing the issue.
